packages/lexical-editor/src/nodes/BaseParagraphNode.ts

```diff
@@ -0,0 +1,190 @@
+        return self;
+    }
+
+    hasTypographyStyle(): boolean {
+        return !!this.getTypographyStyleId();
+    }
+
+    getThemeStyles(): ThemeStyleValue[] {
+        const self = super.getLatest();
+        return self.__styles;
+    }
+
+    setThemeStyles(styles: ThemeStyleValue[]) {
+        const self = super.getWritable();
+        self.__styles = [...styles];
+        return self;
+    }
+
+    static override getType(): string {
+        return "base-paragraph-node";
+    }
+
+    static override clone(node: BaseParagraphNode): BaseParagraphNode {
+        return new BaseParagraphNode(node.getTypographyStyleId(), node.__key);
+    }
+
+    protected updateElementWithThemeClasses(element: HTMLElement, theme: WebinyTheme): HTMLElement {
+        if (!theme?.emotionMap) {
+            return element;
+        }
+
+        if (!this.hasTypographyStyle()) {
+            this.setDefaultTypography(theme.emotionMap);
+        }
+
+        const typoStyleId = this.getTypographyStyleId();
+
+        let themeClasses;
+
+        // Typography css class
+        if (typoStyleId) {
+            const typographyStyle = theme.emotionMap[typoStyleId];
+            if (typographyStyle) {
+                themeClasses = typographyStyle.className;
+            }
+        }
+
+        if (themeClasses) {
+            addClassNamesToElement(element, themeClasses);
+        }
+
+        return element;
+    }
+
+    // View
+    override createDOM(config: EditorConfig): HTMLElement {
+        const element = super.createDOM(config);
+        return this.updateElementWithThemeClasses(element, config.theme as WebinyTheme);
+    }
+
+    override updateDOM(
+        prevNode: BaseParagraphNode,
+        dom: HTMLElement,
+        config: EditorConfig
+    ): boolean {
+        const prevTypoStyleId = prevNode.getTypographyStyleId();
+        const nextTypoStyleId = this.getTypographyStyleId();
+
+        if (!nextTypoStyleId) {
+            this.updateElementWithThemeClasses(dom, config.theme as WebinyTheme);
+            return false;
+        }
+
+        if (prevTypoStyleId !== nextTypoStyleId && nextTypoStyleId) {
+            this.updateElementWithThemeClasses(dom, config.theme as WebinyTheme);
+        }
+        return false;
+    }
+
+    static override importDOM(): DOMConversionMap | null {
+        return {
+            p: () => ({
+                conversion: convertParagraphElement,
+                priority: 0
+            })
+        };
+    }
+
+    static override importJSON(serializedNode: SerializeBaseParagraphNode): BaseParagraphNode {
+        const node = $createBaseParagraphNode();
+        node.setFormat(serializedNode.format);
+        node.setIndent(serializedNode.indent);
+        node.setDirection(serializedNode.direction);
+        node.setThemeStyles(serializedNode.styles);
+        return node;
+    }
+
+    override exportJSON(): SerializeBaseParagraphNode {
+        return {
+            ...super.exportJSON(),
```

Comment:
Not a single comment in this file. 😕

Let's add some comments.

Most importantly, I don't understand why we need BaseParagraph that extends the one from Lexical. I read your Slack messages, but didn't still understand what's going on.

I'm not saying we need to comment on every single method and line, but for starters, let's have a summary at the top of the file, explaining what are the key methods and what they do.

Thanks!